### PR TITLE
:sparkles:Add a work queue for logicalclusters informer.

### PR DESCRIPTION
**Pull request by andreyod**
## Summary
Add a work queue for multi-cluster client logicalclusters informer.
 
## Related issue(s)

Fixes #

### CODE REVIEW

File: pkg/mcclient/controller.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+func (c *controller) handleAdd(cluster interface{}) {
+	clusterInfo, ok := cluster.(*v1alpha1.LogicalCluster)
+	if !ok {
+		runtime.HandleError(errors.New("unexpected object type. expected LogicalCluster"))
+		return
+	}
+	// only clusters in ready state are cached
+	if clusterInfo.Status.Phase != v1alpha1.LogicalClusterPhaseReady {
+		c.handleDelete(clusterInfo.Name)
```
Comment (ezrasilvera) [inline pkg/mcclient/controller.go:150]: Why delete ?  Don't we need to just wait till Ready, e.g., ignore this and wait for additional status change

File: pkg/mcclient/mcclient.go [hunk 2]
```diff
@@ -21,46 +21,47 @@ package mcclient
...
-	// CrossClusterListWatch returns cross-cluster ListWatch
-	CrossClusterListWatch(gv schema.GroupVersion, resource string, namespace string, fieldSelector fields.Selector) *listwatch.CrossClusterListerWatcher
 }
 
 type multiClusterClient struct {
-	configs    map[string]*rest.Config
-	clientsets map[string]*mcclientset.Clientset
-	lock       sync.Mutex
+	ctx              context.Context
+	configs          map[string]*rest.Config
```
Comment (ezrasilvera) [inline pkg/mcclient/mcclient.go:48]: Can't we get the config from the clientset in case someone needs it?

File: pkg/mcclient/mcclient.go [hunk 3]
```diff
@@ -21,46 +21,47 @@ package mcclient
...
 
 func (mcc *multiClusterClient) Cluster(name string) mcclientset.Interface {
 	mcc.lock.Lock()
 	defer mcc.lock.Unlock()
 	clientset, ok := mcc.clientsets[name]
 	if !ok {
-		//TODO change to ClusterOrDie and panic? return an error?
-		panic("invalid cluster name")
+		// try to get LogicalCluster from API server
+		clientset, err := mcc.getFromServer(name)
```
Comment (ezrasilvera) [inline pkg/mcclient/mcclient.go:60]: Why don't we cache the config & cs when doing getFromServer ?  If we missed somehow the event you will come here every time someone use this cluster, right ?

File: pkg/mcclient/controller.go (hunk 1 above)
Comment (andreyod) [reply pkg/mcclient/controller.go:150]: It only deletes from the cache maps. If `Status.Phase` is not ready then the cluster is not accessible and config/cs is not valid. Client code which try to use this config/cs will fail. We will get another event when the cluster is Ready and then we cache it.

File: pkg/mcclient/mcclient.go (hunk 3 above)
Comment (andreyod) [reply pkg/mcclient/mcclient.go:60]: We do cache it in `getFromServer ` function if it exists and valid.

File: pkg/mcclient/mcclient.go (hunk 2 above)
Comment (andreyod) [reply pkg/mcclient/mcclient.go:48]: I don't see an easy way to do this.

File: pkg/mcclient/controller.go (hunk 1 above)
Comment (ezrasilvera) [reply pkg/mcclient/controller.go:150]: Thanks. I forgot you use the same function for Update as well. Can you please add a comment (copy from your reply :-) ) ?

File: pkg/mcclient/mcclient.go (hunk 3 above)
Comment (ezrasilvera) [reply pkg/mcclient/mcclient.go:60]: right, the gitdiff confused me with  the ondelete change :(

File: pkg/mcclient/mcclient.go [hunk 4]
```diff
@@ -21,46 +21,47 @@ package mcclient
...
+		if err == nil {
+			return clientset
```
Comment (ezrasilvera) [inline pkg/mcclient/mcclient.go:62]: nit - if you negate the condition and panik in it, you can remove the return clientset from here

File: pkg/mcclient/mcclient.go [hunk 5]
```diff
@@ -21,46 +21,47 @@ package mcclient
...
-	lock       sync.Mutex
+	ctx              context.Context
+	configs          map[string]*rest.Config
+	clientsets       map[string]*mcclientset.Clientset
+	managerClientset *ksclientset.Clientset
+	lock             sync.Mutex
 }
 
 func (mcc *multiClusterClient) Cluster(name string) mcclientset.Interface {
 	mcc.lock.Lock()
```
Comment (ezrasilvera) [inline pkg/mcclient/mcclient.go:55]: I'm little not at ease calling getFromServer from within a lock.  Also it's too wide. 
If we change indeed there might be races between the explicit Get and the events. 
Can you please at minimal add a comment on getFromServer saying it must be called within an mcc lock ?

File: pkg/mcclient/controller.go (hunk 1 above)
Comment (andreyod) [reply pkg/mcclient/controller.go:151]: Done.

File: pkg/mcclient/mcclient.go (hunk 4 above)
Comment (andreyod) [reply pkg/mcclient/mcclient.go]: Changed.

File: pkg/mcclient/mcclient.go (hunk 5 above)
Comment (andreyod) [reply pkg/mcclient/mcclient.go:56]: Comment added.

**Pull request merged**